File: Go/GoSampleProject/src/etcd/etcd_client_example.go
package etcd

import (
	"context"
	"github.com/google/uuid"
	clientv3 "go.etcd.io/etcd/client/v3"
	"my.sample/GoSampleProject/src/log"
	"time"
)

var PREFIX = "/GoKey"

func EtcdClient() {
	ctx := context.Background()

	// 建立连接
	endpoint := "http://localhost:2379"
	var conn_timeout = 1
	etcd, err := clientv3.New(clientv3.Config{
		//Endpoints:   []string{"localhost:2379", "localhost:22379", "localhost:32379"},
		Endpoints:   []string{endpoint},
		DialTimeout: time.Duration(conn_timeout) * time.Second,
	})
	if err != nil {
		log.Logger.Error().Stack().Err(err).Msg("Etcd连接失败")
		panic(err)
	}
	// 创建一个具有指定超时时间的 context
	timoutCtx, cancel := context.WithTimeout(context.Background(), time.Duration(conn_timeout)*time.Second)
	_, err = etcd.Status(timoutCtx, endpoint)
	cancel()
	if err != nil {
		log.Logger.Error().Stack().Err(err).Msg("Etcd连接失败")
		panic(err)
	}

	log.Logger.Info().Msg("Etcd连接成功")

	// 程序关闭前，释放连接
	defer func() {
		if etcd != nil {
			log.Logger.Info().Msg("关闭Etcd连接")
			etcd.Close()
		}
	}()

	// 设定Key
	key1 := PREFIX + "/Key1"
	key2 := PREFIX + "/Key2"
	setValue1 := "使用Go客户端添加_" + uuid.NewString()
	setValue2 := "使用Go客户端添加_" + uuid.NewString()
	_, err = etcd.Put(ctx, key1, setValue1)
	if err != nil {
		log.Logger.Error().Stack().Err(err).Msgf("添加Key[%s]失败", key1)
		panic(err)
	}
	log.Logger.Info().Msgf("添加Key成功    key=%s    value=%s", key1, setValue1)
	_, err = etcd.Put(ctx, key2, setValue2)
	if err != nil {
		log.Logger.Error().Stack().Err(err).Msgf("添加Key[%s]失败", key2)
		panic(err)
	}
	log.Logger.Info().Msgf("添加Key成功    key=%s    value=%s", key2, setValue2)

	// 取得Key
	resp, err := etcd.Get(ctx, key1)
	if err != nil {
		log.Logger.Error().Stack().Err(err).Msgf("取得Key[%s]失败", key1)
		panic(err)
	}
	for _, ev := range resp.Kvs {
		log.Logger.Info().Msgf("取得Key成功    key=%s    value=%s", ev.Key, ev.Value)
	}

	// 取得Key（前缀机制）
	// 前缀查询
	resp, err = etcd.Get(ctx, PREFIX, clientv3.WithPrefix())
	if err != nil {
		log.Logger.Error().Stack().Err(err).Msgf("取得前缀Key[%s]失败", key1)
		panic(err)
	}
	log.Logger.Info().Msgf("取得前缀成功    prefix=%s", PREFIX)
	for _, ev := range resp.Kvs {
		log.Logger.Info().Msgf("key=%s    value=%s", ev.Key, ev.Value)
	}

	// 分布式锁
	// 官方还没有给Lock的接口，所以需要自己实现

	// 调用租约函数
	leaseRegister(ctx, etcd)

	// 调用监听函数
	watchService(ctx, etcd)

	// 在此堵塞线程
	time.Sleep(5000 * time.Second)
}

// 租约函数
func leaseRegister(ctx context.Context, etcd *clientv3.Client) {
	// 创建租约
	leaseKey := PREFIX + "/LsKey1"
	leaseValue := "此Key会自动被删除"
	var ttl int64 = 5
	resp, err := etcd.Grant(ctx, ttl)
	if err != nil {
		log.Logger.Error().Stack().Err(err).Msgf("创建租约(Lease)失败")
		panic(err)
	}
	log.Logger.Info().Msgf("创建租约(Lease)成功    leaseId=%d", resp.ID)
	_, err = etcd.Put(ctx, leaseKey, leaseValue, clientv3.WithLease(resp.ID))
	if err != nil {
		log.Logger.Error().Stack().Err(err).Msgf("添加租约Key成功    key=%s    value=%s    leaseId=%d", leaseKey, leaseValue, resp.ID)
		panic(err)
	}
	log.Logger.Info().Msgf("添加租约Key成功    key=%s    value=%s    leaseId=%d", leaseKey, leaseValue, resp.ID)

	//续租，在租约过期之前续租
	// ch, kaerr := etcd.KeepAlive(ctx, resp.ID)
	// if kaerr != nil {
	// 	log.Logger.Error().Stack().Err(err).Msgf("续租失败")
	// 	panic(kaerr)
	// }
	// ka := <-ch
	// if ka != nil {
	// 	log.Logger.Info().Msgf("ttl:%d", ka.TTL)
	// } else {
	// 	log.Logger.Info().Msgf("Unexpected NULL")
	// }
}

// 监听函数
func watchService(ctx context.Context, etcd *clientv3.Client) {
	watchKey := PREFIX + "/LsKey1"
	rch := etcd.Watch(ctx, watchKey)
	for wresp := range rch {
		for _, ev := range wresp.Events {
			if ev.Type == clientv3.EventTypeDelete {
				log.Logger.Info().Msgf("租约  key=%q 到期，续租", ev.Kv.Key)
				leaseRegister(ctx, etcd)
			}
		}
	}
}

File: Rust/rust_sample/src/etcd_client_example.rs
use etcd_client::{Client, ConnectOptions, Error, EventType, GetOptions, PutOptions};
use std::time::Duration;
use tracing as log;
use uuid::Uuid;

static PREFIX: &'static str = "/RustKey";

struct KV {
    key: String,
    value: String,
}
impl KV {
    #[inline]
    pub fn new(key: impl Into<String>, value: impl Into<String>) -> Self {
        KV {
            key: key.into(),
            value: value.into(),
        }
    }
    #[inline]
    pub fn key(&self) -> &str {
        &self.key
    }
    #[inline]
    pub fn value(&self) -> &str {
        &self.value
    }
}

async fn etcd_client() -> Result<(), Error> {
    // 连接参数
    let conn_timeout: u64 = 1000;
    let option =
        Some(ConnectOptions::new().with_connect_timeout(Duration::from_millis(conn_timeout)));
    // 建立连接
    let mut client = Client::connect(["localhost:2379"], option).await?;
    let resp = client.status().await;
    match resp {
        Ok(resp) => {
            log::info!("Etcd连接成功 Version:{}", resp.version());
        }
        Err(error) => {
            log::error!("Etcd连接失败: {:?}", error);
            panic!();
        }
    };

    // 设定Key
    let key_1: String = format!("{}{}", PREFIX, "/Key1");
    let set_value_1: String = format!("使用Rust客户端添加_{}", Uuid::new_v4().to_string());
    let kv1 = KV::new(key_1, set_value_1);
    let key_2: String = format!("{}{}", PREFIX, "/Key2");
    let set_value_2: String = format!("使用Rust客户端添加_{}", Uuid::new_v4().to_string());
    let kv2 = KV::new(key_2, set_value_2);
    let _ = client.put(kv1.key(), kv1.value(), None).await?;
    log::info!("添加Key成功    key={}    value={}", kv1.key(), kv1.value());
    let _ = client.put(kv2.key(), kv2.value(), None).await?;
    log::info!("添加Key成功    key={}    value={}", kv2.key(), kv2.value());

    // 取得Key
    let resp = client.get(kv1.key(), None).await?;
    if let Some(kv) = resp.kvs().first() {
        log::info!(
            "取得Key成功    key={}    value={}",
            kv.key_str()?,
            kv.value_str()?
        );
    }

    // 取得Key（前缀机制）
    // 前缀查询
    let resp = client
        .get(PREFIX, Some(GetOptions::new().with_prefix()))
        .await?;
    log::info!("取得前缀成功    prefix={}", PREFIX);
    for kv in resp.kvs() {
        log::info!("key={}    value={}", kv.key_str()?, kv.value_str()?);
    }

    // 分布式锁
    let lock_key: String = format!("{}{}", PREFIX, "/lock");
    let resp = client.lock(lock_key.clone(), None).await?;
    let lock_resp_key = resp.key();
    let lock_resp_key_utf8 = std::str::from_utf8(lock_resp_key)?;
    log::info!(
        "[加锁]分布式锁成功    lockKey={}    lockResKey={}",
        lock_key,
        lock_resp_key_utf8
    );
    log::info!("5秒后解锁,此期间运行 [etcdctl lock /RustKey/lock] 会一直等待");
    tokio::time::sleep(Duration::from_secs(5)).await;
    client.unlock(lock_resp_key).await?;
    log::info!("[解锁]分布式锁成功    lockResKey={}", lock_resp_key_utf8);

    // 调用租约函数
    let _ = lease_register(&mut client).await?;

    // 调用监听函数
    let _ = watch_service(&mut client).await?;

    Ok(())
}

// 租约函数
async fn lease_register(client: &mut Client) -> Result<(), Error> {
    // 创建租约
    let ttl: i64 = 5;
    let resp = client.lease_grant(ttl, None).await?;
    let lease_id = resp.id();
    log::info!("创建租约(Lease)成功    leaseId={}", lease_id);
    // 写入key-value值时绑定一个租约，租约到期后，key会被删除
    let lease_key: String = format!("{}{}", PREFIX, "/LsKey1");
    let lease_value: String = String::from("此Key会自动被删除");
    let lease_kv = KV::new(lease_key, lease_value);
    let _ = client
        .put(
            lease_kv.key(),
            lease_kv.value(),
            Some(PutOptions::new().with_lease(lease_id)),
        )
        .await?;
    log::info!(
        "添加租约Key成功    key={}    value={}    leaseId={}",
        lease_kv.key(),
        lease_kv.value(),
        lease_id
    );

    // 持续续约
    // client.lease_keep_alive(lease_id).await?;

    Ok(())
}

// 监听函数
async fn watch_service(client: &mut Client) -> Result<(), Error> {
    let watch_key: String = format!("{}{}", PREFIX, "/LsKey1");
    let (watcher, mut stream) = client.watch(watch_key, None).await?;
    log::info!("添加监听器成功 watch_id:{}", watcher.watch_id());
    while let Some(resp) = stream.message().await? {
        for event in resp.events() {
            if EventType::Delete == event.event_type() {
                if let Some(kv) = event.kv() {
                    log::info!("租约  key={} 到期，续租", kv.key_str()?);
                }
                let _ = lease_register(client).await;
            }
        }
    }
    Ok(())
}

#[tokio::main]
async fn main() -> Result<(), Error> {
    // install global collector configured based on RUST_LOG env var.
    tracing_subscriber::fmt::init();

    let _ = etcd_client().await?;

    Ok(())
}
